pageserver/src/tenant/timeline.rs

```diff
@@ -1111,16 +1110,6 @@ impl Timeline {
         layers_to_evict: &[Layer],
         cancel: &CancellationToken,
     ) -> anyhow::Result<Vec<Option<Result<(), EvictionError>>>> {
-        // ensure that the layers have finished uploading
-        // (don't hold the layer_removal_cs while we do it, we're not removing anything yet)
-        remote_client
-            .wait_completion()
-            .await
-            .context("wait for layer upload ops to complete")?;
-
-        // now lock out layer removal (compaction, gc, timeline deletion)
-        let _layer_removal_guard = self.layer_removal_cs.lock().await;
```

Comment:
So I guess we would now block on `evict_and_wait` if a compaction was going on and touched some of the layers we're about to evict, right?

This may be a bit OT for this PR, but it makes me wonder if we should make the wanted_evicted field hold a timestamp (the access time when eviction decided to evict the layer), and if the layer has been accessed more recently then we shouldn't go through with  the eviction.  That's obviously an optimization thing rather than a safety thing though.

Reply:
> This may be a bit OT for this PR, but it makes me wonder if we should make the wanted_evicted field hold a timestamp (the access time when eviction decided to evict the layer), and if the layer has been accessed more recently then we shouldn't go through with the eviction. That's obviously an optimization thing rather than a safety thing though.

I did consider that but it is difficult to say would this help. This would move the decision to evict or not to Layer, which I am not keen on doing because you cannot have a full picture there. Theoretically the value is already there. I could add a warning metric for when eviction happens to within one minute accessed Layer.

Currently evictions always lose to re-downloads, meaning that we could have:
1. Layer::evict_and_wait
2. last `Arc<DownloadedLayer>` is dropped, `spawn_blocking` is scheduled
3. before `spawn_blocking` gets to run, there's a new read access to layer contents which gets to run first incrementing the version -- it finds the file locally
4. when spawn_blocking finally gets to run, we notice the version mismatch and eviction is cancelled.

Could be the step 4 is actually a bit racy now that I think of it. Need to check that as well. However this is as designed and as I would like to keep it. Perhaps I just need to get rid of the version field and the "rare metric".

Reply:
To the original:
> So I guess we would now block on evict_and_wait if a compaction was going on and touched some of the layers we're about to evict, right?

No blocking needed. We would just evict or fail to evict regardless of future compaction access (which would start by first acquiring RAII guards keeping the layers resident -- `ResidentLayer`).

Reply:
Found a definitive forever busy loop possibility in case the spawn_blocking never runs, and before that it will be a busy loop.

Reply:
Coming back to this after #5742, #5743, #5744, #5756, #5766, which were pending anyways.

Reply:
This question was about wanted_evicted (enum variant and an atomic); it has now been implemented as eviction cancelling on access; we have non-zero occurances of those in production. Resolving.